UnitystationLauncher/Services/CodeScanService.cs

```diff
@@ -0,0 +1,357 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Threading.Tasks;
+using Serilog;
+using UnitystationLauncher.Models.Enums;
+using UnitystationLauncher.Services.Interface;
+
+namespace UnitystationLauncher.Services;
+
+public class CodeScanService : ICodeScanService
+{
+    private readonly IAssemblyChecker _IAssemblyChecker;
+    private readonly IEnvironmentService _environmentService;
+    private readonly IGoodFileService _iGoodFileService;
+    
+    
+    private const string Managed = "Managed";
+    private const string Plugins = "Plugins";
+
+    
+    
+    
+    public CodeScanService(IAssemblyChecker assemblyChecker, IEnvironmentService environmentService, IGoodFileService iGoodFileService)
+    {
+        _IAssemblyChecker = assemblyChecker;
+        _environmentService = environmentService;
+        _iGoodFileService = iGoodFileService;
+    }
+    
+    
+
+    class FileInfoComparer : IEqualityComparer<FileInfo>
+    {
+        public bool Equals(FileInfo? x, FileInfo? y)
+        {
+            if (x == null || y == null) return false;
+            return x.Name.Equals(y.Name, StringComparison.OrdinalIgnoreCase);
+        }
+
+        public int GetHashCode(FileInfo obj)
+        {
+            return obj.Name.GetHashCode();
+        }
+    }
+
+    public async Task<bool> OnScan( ZipArchive archive, string targetDirectory, string goodFileVersion, Action<string> info, Action<string> errors)
+    {
+        // TODO: Enable extraction cancelling
+        var root = new DirectoryInfo( _environmentService.GetUserdataDirectory());
+        
+        DirectoryInfo stagingDirectory = root.CreateSubdirectory("UnsafeBuildZipDirectory"); 
+        DirectoryInfo processingDirectory = root.CreateSubdirectory("UnsafeBuildProcessing");
+        DirectoryInfo? dataPath = null;
+        archive.ExtractToDirectory(stagingDirectory.ToString(), true);
+        try
+        {
+            DeleteContentsOfDirectory(processingDirectory);
+            info.Invoke("Copying files");
+            CopyFilesRecursively(stagingDirectory.ToString(), processingDirectory.ToString());
+
+            info.Invoke("Cleaning out Dlls and Executables");
+            DeleteFilesWithExtension(processingDirectory.ToString(), "exe");
+            DeleteFilesWithExtension(processingDirectory.ToString(), "dll");
+            DeleteFilesWithExtension(processingDirectory.ToString(), ".so");
+            DeleteFilesWithExtension(processingDirectory.ToString(), ".dylib");
+            if (_environmentService.GetCurrentEnvironment() == CurrentEnvironment.MacOsStandalone)
+            {
+                DeleteFilesWithExtension(processingDirectory.ToString(), ".bundle", exceptionDirectory: Path.Combine(processingDirectory.ToString(), @"Contents\Resources\Data\StreamingAssets")  );
+            }
```

Comment:
I think this would miss the Linux executable, it has no file extension. Not quite sure how you would detect this though, since its possible to change the name from `Unitystation` or include any new ones there too.

![image](https://github.com/unitystation/stationhub/assets/5573038/01025c68-c3f2-46f1-9c7e-998c2a6204f5)

Reply:
maybe you could check by the executable flag?

Reply:
Yeah that could work, it might miss things sometimes if someone does a `chmod -x` on it though.

Reply:
that should be fine since that would be something the user does manually?
though I don't know if unity does it automatically in any circumstance?

Reply:
For some reason github thinks I really need to type into this box

Reply:
nope help it keeps on redirecting me to this box  when I try looking at the code xD